test/parallel_api/ranges/std_ranges_test.h

```diff
@@ -0,0 +1,455 @@
+#include <span>
+#include <iostream>
+#include <vector>
+#include <typeinfo>
+#include <string>
+
+namespace test_std_ranges
+{;
+
+#if _ONEDPL_HETERO_BACKEND
+template<int call_id = 0>
+auto dpcpp_policy()
+{
+    auto exec = TestUtils::default_dpcpp_policy;
+    using Policy = decltype(exec);
+    return TestUtils::make_new_policy<TestUtils::new_kernel_name<Policy, call_id>>(TestUtils::default_dpcpp_policy);
+}
+#endif //_ONEDPL_HETERO_BACKEND
+
+auto host_policies() { return std::true_type{};}
+
+enum TestDataMode
+{
+    data_in,
+    data_in_out,
+    data_in_in,
+    data_in_in_out
+};
+
+auto f_mutuable = [](auto&& val) { return val *= val; };
+auto proj_mutuable = [](auto&& val) { return val *= 2; };
+
+auto f = [](auto&& val) { return val * val; };
+auto proj = [](auto&& val){ return val * 2; };
+auto pred = [](auto&& val) { return val == 5; };
+auto pred_2 = [](auto&& val1, auto&& val2) { return val1 == val2; };
+
+auto pred1 = [](auto&& val) -> decltype(auto) { return val > 0; };
+auto pred2 = [](auto&& val) -> decltype(auto) { return val == 4; };
+auto pred3 = [](auto&& val) -> decltype(auto) { return val < 0; };
+
+struct P2
+{
+    P2(auto v): x(v) {}
+    int x = {};
+    int y = {};
+
+    int proj() { return x; }
+    friend bool operator==(const P2& a, const P2& b) { return a.x == b.x && a.y == b.y; }
+};
+
+template<typename DataType, typename Container, TestDataMode Ranges = data_in, bool RetTypeCheck = true>
+struct test
+{
+    template<typename Policy, typename Algo, typename... Args>
+    std::enable_if_t<std::is_same_v<Policy, std::true_type>>
+    operator()(Policy, Algo algo, Args... args)
+    {
+        operator()(oneapi::dpl::execution::seq, algo, args...);
+        operator()(oneapi::dpl::execution::unseq, algo, args...);
+        operator()(oneapi::dpl::execution::par, algo, args...);
+        operator()(oneapi::dpl::execution::par_unseq, algo, args...);
+    }
+
+    template<typename Policy, typename Algo, typename Checker, typename Transform>
+    std::enable_if_t<!std::is_same_v<Policy, std::true_type> && Ranges == data_in>
+    operator()(Policy&& exec, Algo algo, Checker checker, Transform tr, auto... args)
+    {
+        constexpr int max_n = 10;
+        DataType data[max_n] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+        DataType expected[max_n] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+
+        auto expected_view = tr(std::ranges::subrange(expected, expected + max_n));
+        auto expected_res = checker(expected_view, args...);
+        {
+            Container cont(exec, std::ranges::begin(data), max_n);
+            typename Container::type& A = cont();
+
+            auto res = algo(exec, tr(A), args...);
+
+            //check result
+            if constexpr(RetTypeCheck)
+                static_assert(std::is_same_v<decltype(res), decltype(checker(tr(A), args...))>, "Wrong return type");
+
+            auto bres = ret_in_val(expected_res, expected_view.begin()) == ret_in_val(res, tr(A).begin());
```

Comment:
Could you describe the purpose if this function and what happens here? 
It is pretty difficult to understand due to names like `reit_in_val`, `in`, `in1`, `in2`, `check_in`, `check_in1`, which are very elaborate...

Reply:
I think I understand it now. `in`, `in1`, `in2` and `out`  data members used by `ret_in_val` are from [std::ranges::in_out_result](https://en.cppreference.com/w/cpp/algorithm/ranges/return_types/in_out_result) and [std::ranges::in_in_out_result](https://en.cppreference.com/w/cpp/algorithm/ranges/return_types/in_in_out_result), which are used by [transform](https://en.cppreference.com/w/cpp/algorithm/ranges/transform). This was the major thing, which puzzled me.
`ret_in_val` retrieves a distance between the beginning of a sequence and an abstract object returned by an algorithm.